File: daemon/internal/protocol/agents.go
package protocol

import "time"

// AgentModel is one model that an agent can run. The pickers list it, and ID is the value that
// goes back to the daemon when a card is created.
type AgentModel struct {
	// ID is the value the agent takes as its model setting, such as "sonnet" or "gemini-2.5-pro".
	ID string `json:"id"`
	// Name is the words shown to people.
	Name string `json:"name"`
	// Thinking says whether the model has a thinking setting at all. The thinking picker is only
	// offered when it is true and the agent's own Thinking capability is true as well, because
	// an agent may have models that think and still give Marshal no way to set how hard.
	Thinking bool `json:"thinking"`
}

// AgentCapabilities says what Marshal can do with an agent, as far as it is known before a
// session starts. It describes the agent as Marshal drives it, so a feature the agent has but
// Marshal does not use yet is false.
type AgentCapabilities struct {
	// Resume is true when a stopped session can be picked up again with its session id.
	Resume bool `json:"resume"`
	// StructuredEvents is true when the agent reports messages and tool calls as events, so the
	// chat view works. An agent without it is only shown as a terminal.
	StructuredEvents bool `json:"structuredEvents"`
	// ModelSwitching is true when a card can choose the model when its session starts.
	ModelSwitching bool `json:"modelSwitching"`
	// Thinking is true when a card can choose a thinking mode when its session starts.
	Thinking bool `json:"thinking"`
	// MCP is true when the agent accepts MCP servers from Marshal.
	MCP bool `json:"mcp"`
	// Approvals is true when the agent can ask the person before it uses a tool. When it is
	// false, the "Ask" permission mode makes the agent skip what needs approval instead of
	// asking, and the screens should say so.
	Approvals bool `json:"approvals"`
}

// Agent is one entry of the agent catalog: a kind of coding agent, and whether it can be used on
// this machine. The place of the program on disk is not sent, because it names the person's
// folders.
type Agent struct {
	// Kind says which agent it is.
	Kind AgentKind `json:"kind"`
	// Name is the words shown to people, such as "Claude Code".
	Name string `json:"name"`
	// Version is the installed version, or empty when the agent is missing.
	Version string `json:"version"`
	// Status says whether the agent can be used here.
	Status AgentStatus `json:"status"`
	// Warning is one plain sentence for the screens when there is something to know, such as an
	// untested version. It is empty when there is nothing to say.
	Warning string `json:"warning"`
	// InstallHint is one plain sentence with the install command, for the disabled row of a
	// missing agent. It is empty when the agent is installed.
	InstallHint string `json:"installHint"`
	// Models are the models the agent can run, in the order to show them. The first one is the
	// default when a card switches to this agent.
	Models []AgentModel `json:"models"`
	// Capabilities says what Marshal can do with the agent.
	Capabilities AgentCapabilities `json:"capabilities"`
}

// AgentCatalog is the answer to GET /v1/agents.
type AgentCatalog struct {
	// Agents has one entry for each kind of agent that a card can use, whether or not it is
	// installed. The built-in agent is not listed here.
	Agents []Agent `json:"agents"`
	// ServerTime is the daemon's time when the answer was made. The catalog itself may come from
	// a check that was made a few minutes earlier.
	ServerTime Timestamp `json:"serverTime"`
}

// NewAgentCatalog makes a catalog stamped with the daemon's time. It turns a nil list into an
// empty one, for the catalog and for each agent's models, so the JSON has [] and never null.
func NewAgentCatalog(agents []Agent, now time.Time) AgentCatalog {
	out := make([]Agent, len(agents))
	for i, agent := range agents {
		if agent.Models == nil {
			agent.Models = []AgentModel{}
		}
		out[i] = agent
	}
	return AgentCatalog{Agents: out, ServerTime: NewTimestamp(now)}
}

File: apps/web/src/data/mappers/agents.ts
import type {
  Agent,
  AgentCapabilities,
  AgentCatalog,
  AgentKind,
  AgentModel,
  AgentStatus,
} from "@marshal/protocol";

/** One entry of the agent picker. An agent that is not installed is here too, marked as missing. */
export interface AgentOption {
  kind: AgentKind;
  /** The words shown to people, such as "Claude Code". */
  name: string;
  version: string;
  status: AgentStatus;
  /** True when the agent is not installed. The row is shown but cannot be picked. */
  missing: boolean;
  /** One plain sentence when there is something to know, such as an untested version. Else empty. */
  warning: string;
  /** One plain sentence with the install command, for a missing agent. Else empty. */
  installHint: string;
  models: AgentModel[];
  capabilities: AgentCapabilities;
}

/**
 * One agent in the shape of the prototype's `AGENTS` table, which the pickers and the card
 * settings read. It lives in the data layer, not in `mock/`, so the mapper never depends on the
 * mock; `mock/types.ts` re-exports it for the store's own use.
 */
export interface AgentInfo {
  models: string[];
  icon: string;
  version: string;
}

/** What the mappers read of the catalog: its agents. The daemon's answer has the server time too, and the store keeps only the agents. */
export type CatalogAgents = Pick<AgentCatalog, "agents"> &
  Partial<Pick<AgentCatalog, "serverTime">>;

/* The icons the mock's `AGENTS` table uses, so the pickers look the same on either side. */
const AGENT_ICONS: Record<AgentKind, string> = {
  claude: "terminal-square",
  gemini: "terminal-square",
  codex: "terminal-square",
  builtin: "cpu",
};

function toOption(agent: Agent): AgentOption {
  return {
    kind: agent.kind,
    name: agent.name,
    version: agent.version,
    status: agent.status,
    missing: agent.status === "missing",
    warning: agent.warning,
    installHint: agent.installHint,
    models: agent.models.map((model) => ({ ...model })),
    capabilities: { ...agent.capabilities },
  };
}

/** The catalog as a list for the pickers, in the order the daemon gives, with no agent dropped. */
export function toAgentOptions(catalog: CatalogAgents): AgentOption[] {
  return catalog.agents.map(toOption);
}

/**
 * The catalog in the shape of the mock's `AGENTS`: keyed by the name shown to people, with the
 * model ids (what a card sends back to the daemon), an icon, and the version. The built-in agent
 * is not in the daemon's catalog, so it has no key here. A missing agent has one, with an empty version.
 */
export function toLegacyAgents(catalog: CatalogAgents): Record<string, AgentInfo> {
  return Object.fromEntries(
    catalog.agents.map((agent) => [
      agent.name,
      {
        models: agent.models.map((model) => model.id),
        icon: AGENT_ICONS[agent.kind],
        version: agent.version,
      },
    ]),
  );
}

/**
 * A function like the mock's `thinkSupported`: does the thinking picker apply to this model? The
 * model must have a thinking setting, and the agent must let Marshal set it, because an agent may
 * have models that think and still give Marshal no way to say how hard (`AgentModel.Thinking` in
 * the Go types). A model the catalog does not list has none.
 */
export function thinkSupportedIn(catalog: CatalogAgents): (model: string) => boolean {
  const supported = new Set(
    catalog.agents
      .filter((agent) => agent.capabilities.thinking)
      .flatMap((agent) => agent.models.filter((model) => model.thinking).map((model) => model.id)),
  );
  return (model) => supported.has(model);
}
